language/move-prover/bytecode-to-boogie/src/bytecode_instrs.bpl

```diff
@@ -15,16 +15,67 @@ function {:constructor} Path(p: [int]Edge, size: int): Path;
 
 const DefaultPath: [int]Edge;
 
+type {:datatype} TypeValue;
+function {:constructor} BooleanType() : TypeValue;
+function {:constructor} IntegerType() : TypeValue;
+function {:constructor} AddressType() : TypeValue;
+function {:constructor} ByteArrayType() : TypeValue;
+function {:constructor} StrType() : TypeValue;
+function {:constructor} ReferenceType(t: TypeValue) : TypeValue;
+function {:constructor} VectorType(t: TypeValue) : TypeValue;
+// StructTypes caries the arity (# of fields) so we can enumerate fields (which are ints)
+function {:constructor} StructType(name: TypeName, arity: int, ts: [FieldName]TypeValue) : TypeValue;
+
 type {:datatype} Value;
 function {:constructor} Boolean(b: bool): Value;
 function {:constructor} Integer(i: int): Value;
 function {:constructor} Address(a: Address): Value;
 function {:constructor} ByteArray(b: ByteArray): Value;
 function {:constructor} Str(a: String): Value;
-function {:constructor} Map(m: [Edge]Value): Value;
-function {:constructor} Vector(v: [Edge]Value, l: int): Value;
+// Currently empty vectors of arbitrary type will be identified as equal. If this causes a problem,
+// we need to add the element's TypeValue to the vector.
+function {:constructor} Vector(v: [int]Value, l: int): Value;
+// Currently structural equal structs are identified. We need to add the type name below if this causes a problem.
+function {:constructor} Struct(v: [FieldName]Value): Value;
+
+// Axiomatic definition of type test and equality.
+function has_type(t: TypeValue, v: Value): bool;
+function is_equal(t: TypeValue, v1: Value, v2: Value): bool;
+
+procedure {:inline 1} is_equal_proc(t: TypeValue, v1: Value, v2: Value) returns (r: Value);
+  ensures r == Boolean(is_equal(t, v1, v2));
+
+procedure {:inline 1} is_not_equal_proc(t: TypeValue, v1: Value, v2: Value) returns (r: Value);
+  ensures r == Boolean(!is_equal(t, v1, v2));
+
+axiom (forall x: bool :: has_type(BooleanType(), Boolean(x)));
+axiom (forall x: int :: has_type(IntegerType(), Integer(x)));
+axiom (forall x: Address :: has_type(AddressType(), Address(x)));
+axiom (forall x: ByteArray :: has_type(ByteArrayType(), ByteArray(x)));
+axiom (forall x: String :: has_type(StrType(), Str(x)));
+axiom (forall x: [int]Value, n: int, t: TypeValue :: has_type(VectorType(t), Vector(x, n)) <==>
+         (forall i: int :: i >= 0 && i < n ==> has_type(t, x[i])));
+axiom (forall name: TypeName, arity: int, ts: [FieldName]TypeValue, v: [FieldName]Value ::
```

Comment:
It looks to me that a Struct {0->0, 1->1} could match multiple types A { int, int}, B {int, int}, etc.  This is because you are not carrying the type name in the value.

Reply:
Perhaps you could leave a TypeName field inside the Struct constructor for Value.

Reply:
I think we should be fine with the fact that a value can have multiple types. The same BTW also applies for Vector (an empty vector can have any vector type). But we can add this if problems appear; I already have a comment for this at the declarations.